File: frontend/lib/server/routing.ts
import type { PharmacyOffer, RouteStop } from "@/lib/mvp/types";

type MinimalPharmacySelection = Pick<
  PharmacyOffer,
  "pharmacyId" | "pharmacyName" | "lat" | "lon"
>;

type UserLocation = {
  lat: number;
  lon: number;
};

type RouteableSelection = Array<Pick<MinimalPharmacySelection, "pharmacyId">>;

export function shouldBuildRoute(pharmacies: RouteableSelection): boolean {
  return pharmacies.length >= 2;
}

export function buildRouteStops(
  userLocation: UserLocation,
  pharmacies: MinimalPharmacySelection[]
): RouteStop[] {
  return [
    {
      pharmacyId: "origin",
      label: "Ваше местоположение",
      lat: userLocation.lat,
      lon: userLocation.lon,
      order: 0
    },
    ...pharmacies.map((pharmacy, index) => ({
      pharmacyId: pharmacy.pharmacyId,
      label: pharmacy.pharmacyName,
      lat: pharmacy.lat,
      lon: pharmacy.lon,
      order: index + 1
    }))
  ];
}

File: backend/app/services/routing.py
from app.schemas import PharmacyOffer, RouteResponse, RouteStop


def build_route(origin: tuple[float, float], pharmacies: list[PharmacyOffer]) -> RouteResponse:
    ordered_stops = [
        RouteStop(
            pharmacy_id="origin",
            label="Ваше местоположение",
            lat=origin[0],
            lon=origin[1],
            order=0,
        ),
        *[
            RouteStop(
                pharmacy_id=pharmacy.pharmacy_id,
                label=pharmacy.pharmacy_name,
                lat=pharmacy.lat,
                lon=pharmacy.lon,
                order=index + 1,
            )
            for index, pharmacy in enumerate(pharmacies)
        ],
    ]
    return RouteResponse(
        total_duration_minutes=12 + len(pharmacies) * 3,
        total_distance_km=round(2.4 + len(pharmacies) * 1.1, 1),
        ordered_stops=ordered_stops,
    )
